README.md
# react-native-udp

React Native UDP socket API for Android & iOS. It allows you to create UDP sockets, imitating Node's [dgram](https://nodejs.org/api/dgram.html) API functionalities (check each method documentation for more information).

_This module is used by [Tradle](https://github.com/tradle)._

## Table of Contents

- [Getting started](#getting-started)
- [Compatibility](#react-native-compatibility)
- [Usage](#usage)
- [Maintainers](#maintainers)
- [License](#license)

## Getting started

Install the library using either Yarn:

```
yarn add react-native-udp
```

or npm:

```
npm install --save react-native-udp
```

#### Using React Native >= 0.60

Linking the package manually is not required anymore with [Autolinking](https://github.com/react-native-community/cli/blob/master/docs/autolinking.md).

- **iOS Platform:**

  `$ cd ios && pod install && cd ..` # CocoaPods on iOS needs this extra step

#### Using React Native < 0.60

You then need to link the native parts of the library for the platforms you are using. The easiest way to link the library is using the CLI tool by running this command from the root of your project:

`$ react-native link react-native-udp`

If you can't or don't want to use the CLI tool, you can also manually link the library using the instructions below (click on the arrow to show them):

<details>
<summary>Manually link the library on iOS</summary>

1. In XCode, in the project navigator, right click `Libraries` ➜ `Add Files to [your project's name]`
2. Go to `node_modules` ➜ `react-native-udp` and add `UdpSockets.xcodeproj`
3. In XCode, in the project navigator, select your project. Add `libUdpSockets.a` to your project's `Build Phases` ➜ `Link Binary With Libraries`
4. Run your project (`Cmd+R`)<
   </details>

<details>
<summary>Manually link the library on Android</summary>

1. Open up `android/app/src/main/java/[...]/MainApplication.java`

- Add `import com.tradle.react.UdpSocketsModule;` to the imports at the top of the file
- Add `new UdpSocketsModule()` to the list returned by the `getPackages()` method

2. Append the following lines to `android/settings.gradle`:
   ```
   include ':react-native-udp'
   project(':react-native-udp').projectDir = new File(rootProject.projectDir, 	'../node_modules/react-native-udp/android')
   ```
3. Insert the following lines inside the dependencies block in `android/app/build.gradle`:
   `compile project(':react-native-udp')`
   </details>

## React Native Compatibility

| React Native Version | Use Version |
| -------------------- | ----------- |
| <=0.59.10            | <=2.7.0     |
| >=0.60.0             | >=3.0.0     |

## Usage

_see/run [index.js](examples/udpsockets) for a complete example, but the interface is like dgram's_

```js
import dgram from 'react-native-udp'

const socket = dgram.createSocket('udp4')
socket.bind(12345)
socket.once('listening', function() {
  socket.send('Hello World!', undefined, undefined, remotePort, remoteHost, function(err) {
    if (err) throw err

    console.log('Message sent!')
  })
})

socket.on('message', function(msg, rinfo) {
  console.log('Message received', msg)
})
```

### Debugging

To log all info emitted from the socket, add `debug` option when creating the socket:

```js
const socket = dgram.createSocket({
  type: 'udp4',
  debug: true,
})
```

## Maintainers

- [Rapsssito](https://github.com/rapsssito) [[Support me :heart:](https://github.com/sponsors/Rapsssito)]
- [Mark Vayngrib](https://github.com/mvayngrib)

## License

The library is released under the MIT license. For more information see [`LICENSE`](/LICENSE).


CHANGELOG.md
## [4.1.7](https://github.com/tradle/react-native-udp/compare/v4.1.6...v4.1.7) (2023-01-26)


### Bug Fixes

* **Android:** Remove incorrect Override signature ([#241](https://github.com/tradle/react-native-udp/issues/241)) ([2578857](https://github.com/tradle/react-native-udp/commit/25788578e59a1704ea65f9d00bc6c41faaae3f3d))

## [4.1.6](https://github.com/tradle/react-native-udp/compare/v4.1.5...v4.1.6) (2022-12-05)


### Bug Fixes

* **Android:** Update Android code ([#220](https://github.com/tradle/react-native-udp/issues/220)) ([e5cb016](https://github.com/tradle/react-native-udp/commit/e5cb016e55133a2683b695b4dfed68cb72252948))

## [4.1.5](https://github.com/tradle/react-native-udp/compare/v4.1.4...v4.1.5) (2022-03-22)


### Bug Fixes

* Fix Gradle configuration ([#205](https://github.com/tradle/react-native-udp/issues/205)) ([c5c7831](https://github.com/tradle/react-native-udp/commit/c5c783188563e9eb04370001e702993b650d6c13))
* **Android:** Catch all exceptions and redirect them to JS ([#202](https://github.com/tradle/react-native-udp/issues/202)) ([0a512ee](https://github.com/tradle/react-native-udp/commit/0a512ee84f7370af1842907c08d79ace29fc2019))

## [4.1.4](https://github.com/tradle/react-native-udp/compare/v4.1.3...v4.1.4) (2022-02-22)


### Bug Fixes

* Remove AsyncTask in the send function (it's deprecated) ([#197](https://github.com/tradle/react-native-udp/issues/197)) ([179a948](https://github.com/tradle/react-native-udp/commit/179a9480004ce15c83fbcaa4ec3d9cd9fa9f950a))

## [4.1.3](https://github.com/tradle/react-native-udp/compare/v4.1.2...v4.1.3) (2021-09-08)


### Bug Fixes

* Fix podspec to be compliant with latest Cocoapods ([#191](https://github.com/tradle/react-native-udp/issues/191)) ([006361b](https://github.com/tradle/react-native-udp/commit/006361b95b1e6c836225cdb95fe733f41bdf2afb))

## [4.1.2](https://github.com/tradle/react-native-udp/compare/v4.1.1...v4.1.2) (2021-01-19)


### Bug Fixes

* **Android:** Fix null MulticastSocket creation ([#148](https://github.com/tradle/react-native-udp/issues/148)) ([3d52791](https://github.com/tradle/react-native-udp/commit/3d527916a24e3837c2eca194f2474af22a737cef))

## [4.1.1](https://github.com/tradle/react-native-udp/compare/v4.1.0...v4.1.1) (2020-12-24)


### Bug Fixes

* Fix TypeScript declaration inheritance ([#153](https://github.com/tradle/react-native-udp/issues/153)) ([3e5b4f7](https://github.com/tradle/react-native-udp/commit/3e5b4f71d021f19d7720b7ccf05a343f93591d15))

# [4.1.0](https://github.com/tradle/react-native-udp/compare/v4.0.4...v4.1.0) (2020-12-05)


### Features

* Add rx timestamp to rinfo payload ([#146](https://github.com/tradle/react-native-udp/issues/146)) ([c468a39](https://github.com/tradle/react-native-udp/commit/c468a3942b7637302b33cf9946f2e89818ea70fb))

## [4.0.4](https://github.com/tradle/react-native-udp/compare/v4.0.3...v4.0.4) (2020-11-28)


### Bug Fixes

* **Android:** Switch to Java Standard concurrent API ([#145](https://github.com/tradle/react-native-udp/issues/145)) ([373edf0](https://github.com/tradle/react-native-udp/commit/373edf03bd274fa8d00609211177bee813978dd1))

## [4.0.3](https://github.com/tradle/react-native-udp/compare/v4.0.2...v4.0.3) (2020-10-01)


### Bug Fixes

* Fix Xcode 12 compatibility ([#138](https://github.com/tradle/react-native-udp/issues/138)) ([750960b](https://github.com/tradle/react-native-udp/commit/750960b8dd7768bd20ea709292e9a4656c9206f4)), closes [facebook/react-native#29633](https://github.com/facebook/react-native/issues/29633)

## [4.0.2](https://github.com/tradle/react-native-udp/compare/v4.0.1...v4.0.2) (2020-09-06)


### Bug Fixes

* Add backwards compatiblity with CommonJS require() ([#137](https://github.com/tradle/react-native-udp/issues/137)) ([af36905](https://github.com/tradle/react-native-udp/commit/af36905bd92dbf78194f2d2d38aa531e9e2b4553))

## [4.0.1](https://github.com/tradle/react-native-udp/compare/v4.0.0...v4.0.1) (2020-09-01)


### Bug Fixes

* Add missing attribute "homepage" to package.json ([#135](https://github.com/tradle/react-native-udp/issues/135)) ([74d75c1](https://github.com/tradle/react-native-udp/commit/74d75c139dbca0049b5c201d5d3144a5c22c09f6))

# [4.0.0](https://github.com/tradle/react-native-udp/compare/v3.2.0...v4.0.0) (2020-08-30)


### Features

* Match NodeJS dgram API ([#133](https://github.com/tradle/react-native-udp/issues/133)) ([2f8dc35](https://github.com/tradle/react-native-udp/commit/2f8dc35d18a7875616bd18d4e6dd5f1d74b6230a)), closes [#132](https://github.com/tradle/react-native-udp/issues/132) [#128](https://github.com/tradle/react-native-udp/issues/128)


### BREAKING CHANGES

* All methods no longer check parameter types. socket.send() now throws 'ERR_SOCKET_BAD_PORT' when sending on unbound sockets. socket.send() no longer checks if ddress is a valid IP or hostname.

# [3.2.0](https://github.com/tradle/react-native-udp/compare/v3.1.0...v3.2.0) (2020-08-15)


### Bug Fixes

* **Android:** createSocket is now a blocking method ([#108](https://github.com/tradle/react-native-udp/issues/108)) ([6796a7f](https://github.com/tradle/react-native-udp/commit/6796a7f12762850262e111a40f84b841ed67c401))


### Features

* Add Typescript types and the option to enable debugging ([#114](https://github.com/tradle/react-native-udp/issues/114)) ([11e6e49](https://github.com/tradle/react-native-udp/commit/11e6e49a417b2a54227977691cd3c5b84a5d5d36))


### Performance Improvements

* Dependency reduction & JS update ([#105](https://github.com/tradle/react-native-udp/issues/105)) ([aac2bfa](https://github.com/tradle/react-native-udp/commit/aac2bfa30591864b18fe8e4190f141f97798e058))


RCTUDPSocket.js
//
//  react-native-udp
//
//  Created by Mark Vayngrib on 05/10/15.
//  Copyright (c) 2015 Tradle, Inc. All rights reserved.
//

/**
 * @providesModule RCTUDPSocket
 * @flow
 */

'use strict';

var React = require('react-native')
var {
  Component
} = React

var mixInEventEmitter = require('mixInEventEmitter')
var DeviceEventEmitter = require('RCTDeviceEventEmitter')
var NativeModules = require('NativeModules')
var sockets = NativeModules.UDP // just UDP for now
var noop = function () {}
var instances = 0
var STATE = {
  UNBOUND: 0,
  BINDING: 1,
  BOUND: 2
}

class RCTSocket extends Component {
  id: String;

  _state: STATE.UNBOUND;

  _address: undefined;

  _port: undefined;

  constructor(props) {
    super(props)
    this.id = instances++
    this.subscriptiom = DeviceEventEmitter.addListener(
      'udp-' + this.id + '-data', this._onReceive.bind(this)
    );

    // ensure compatibility with node's EventEmitter
    if (!this.on) this.on = this.addListener.bind(this)

    sockets.createSocket(this.id, {
      type: props.type || 'udp4'
    }) // later
  }

  _debug() {
    var args = [].slice.call(arguments)
    args.unshift(this.id)
    console.log.apply(console, args)
  }

  bind(port, address, callback) {
    var self = this
    if (typeof address === 'function') {
      callback = address
      address = undefined
    }

    // address = address || '0.0.0.0' //'127.0.0.1'
    if (callback) this.once('listening', callback)

    this._state = STATE.BINDING
    this._debug('binding, address:', address, 'port:', port)
    sockets.bind(this.id, port, address, function(err, addr) {
      if (err) {
        // questionable: may want to self-destruct and
        // force user to create a new socket
        self._state = STATE.UNBOUND
        self._debug('failed to bind', err)
        return self.emit('error', err)
      }

      self._address = addr.address
      self._port = addr.port
      self._state = STATE.BOUND
      self.emit('listening')
    })
  }

  componentWillUnmount() {
    this.subscription.remove();
  }

  _onReceive(info) {
    this._debug('received', info)

    var buf = toByteArray(info.data)
    var rinfo = {
      address: info.address,
      port: info.port,
      family: 'IPv4', // not necessarily
      size: buf.length
    }

    if (typeof Buffer !== 'undefined') buf = new Buffer(buf)

    this.emit('message', buf, rinfo)
  }


  /**
   * socket.send(buf, offset, length, port, address, [callback])
   *
   * For UDP sockets, the destination port and IP address must be
   * specified. A string may be supplied for the address parameter, and it will
   * be resolved with DNS. An optional callback may be specified to detect any
   * DNS errors and when buf may be re-used. Note that DNS lookups will delay
   * the time that a send takes place, at least until the next tick. The only
   * way to know for sure that a send has taken place is to use the callback.
   *
   * If the socket has not been previously bound with a call to bind, it's
   * assigned a random port number and bound to the "all interfaces" address
   * (0.0.0.0 for udp4 sockets, ::0 for udp6 sockets).
   *
   * @param {Array|string} message to be sent
   * @param {number} offset Offset in the buffer where the message starts.
   * @param {number} length Number of bytes in the message.
   * @param {number} port destination port
   * @param {string} address destination IP
   * @param {function} callback Callback when message is done being delivered.
   *                            Optional.
   */
  // Socket.prototype.send = function (buf, host, port, cb) {
  send(buffer, offset, length, port, address, callback) {
    var self = this

    if (offset !== 0) throw new Error('Non-zero offset not supported yet')

    if (this._state === STATE.UNBOUND) {
      throw new Error('bind before sending, seriously dude')
    }
    else if (this._state === STATE.BINDING) {
      // we're ok, GCDAsync(Udp)Socket handles queueing internally
    }

    callback = callback || noop
    if (typeof buffer === 'string') {
      buffer = toByteArray(buffer)
    }
    else if (typeof Buffer !== 'undefined' && Buffer.isBuffer(buffer)) {
      buffer = buffer.toJSON().data
    }

    self._debug('sending', buffer)
    sockets.send(this.id, buffer, +port, address, function(err) {
      if (err) {
        self._debug('send failed', err)
        return callback(err)
      }

      self._debug('sent')
      callback()
    })
  }

  address() {
    if (this._state !== STATE.BOUND) {
      throw new Error('socket is not bound yet')
    }

    return {
      address: this._address,
      port: this._port,
      family: 'IPv4'
    }
  }

  close() {
    var self = this
    if (this._destroyed) return

    this._destroyed = true
    this._debug('closing')
    sockets.close(this.id, function() {
      self._debug('closed')
    })

    this.emit('close')
  }

  setBroadcast(flag) {
    // nothing yet
  }

  setTTL(ttl) {
    // nothing yet
  }

  setMulticastTTL(ttl, callback) {
    // nothing yet
  }

  setMulticastLoopback(flag, callback) {
    // nothing yet
  }

  addMembership(multicastAddress, multicastInterface, callback) {
    // nothing yet
  }

  dropMembership(multicastAddress, multicastInterface, callback) {
    // nothing yet
  }

  ref() {
    // anything?
  }

  unref() {
    // anything?
  }
}

mixInEventEmitter(RCTSocket, {
  'listening': true,
  'message': true,
  'close': true,
  'error': true
})

function toByteArray(obj) {
  if (typeof obj === 'object') {
    var i = 0
    var arr = []
    while (true) {
      if (!(i in obj)) break

      arr.push(+obj[i])
      i++
    }

    return new Uint8Array(arr)
  }
  else if (typeof obj !== 'string') {
    throw new Error('unsupported format')
  }

  var uint = new Uint8Array(obj.length);
  for (var i = 0, l = obj.length; i < l; i++){
    uint[i] = obj.charCodeAt(i);
  }

  return new Uint8Array(uint);
}
module.exports = RCTSocket